examples/text_summarization/pointer_summarizer/src/model.py

```diff
@@ -0,0 +1,233 @@
+                            bias_ih_attr=paddle.ParamAttr(initializer=I.Constant(
+                                value=0.0)))
+
+        if config.pointer_gen:
+            self.p_gen_linear = nn.Linear(config.hidden_dim * 4 + config.emb_dim, 1)
+
+        # p_vocab
+        self.out1 = nn.Linear(config.hidden_dim * 3, config.hidden_dim)
+        self.out2 = nn.Linear(config.hidden_dim, config.vocab_size, weight_attr=paddle.ParamAttr(initializer=I.Normal(
+            std=config.trunc_norm_init_std)))
+
+        print()
+
+    def forward(self, y_t_1, s_t_1, encoder_outputs, encoder_feature, enc_padding_mask,
+                c_t_1, extra_zeros, enc_batch_extend_vocab, coverage, step):
+        if not self.training and step == 0:
+            h_decoder, c_decoder = s_t_1
+            s_t_hat = paddle.concat((paddle.reshape(h_decoder, [-1, config.hidden_dim]),
+                                     paddle.reshape(c_decoder, [-1, config.hidden_dim])), 1)  # B x 2*hidden_dim
+            c_t, _, coverage_next = self.attention_network(s_t_hat, encoder_outputs, encoder_feature,
+                                                           enc_padding_mask, coverage)
+            coverage = coverage_next
+
+        y_t_1_embd = self.embedding(y_t_1)
+        x = self.x_context(paddle.concat((c_t_1, y_t_1_embd), 1))
+        lstm_out, s_t = self.lstm(x.unsqueeze(1), s_t_1)
+
+        h_decoder, c_decoder = s_t
+        s_t_hat = paddle.concat((paddle.reshape(h_decoder, [-1, config.hidden_dim]),
+                                 paddle.reshape(c_decoder, [-1, config.hidden_dim])), 1)  # B x 2*hidden_dim
+        c_t, attn_dist, coverage_next = self.attention_network(s_t_hat, encoder_outputs, encoder_feature,
+                                                               enc_padding_mask, coverage)
+
+        if self.training or step > 0:
+            coverage = coverage_next
+
+        p_gen = None
+        if config.pointer_gen:
+            p_gen_input = paddle.concat((c_t, s_t_hat, x), 1)  # B x (2*2*hidden_dim + emb_dim)
+            p_gen = self.p_gen_linear(p_gen_input)
+            p_gen = F.sigmoid(p_gen)
+
+        output = paddle.concat((paddle.reshape(lstm_out, [-1, config.hidden_dim]), c_t), 1)  # B x hidden_dim * 3
+        output1 = self.out1(output)  # B x hidden_dim
+        output2 = self.out2(output1)  # B x vocab_size
+        vocab_dist = F.softmax(output2, axis=1)
+
+        if config.pointer_gen:
+            vocab_dist_ = p_gen * vocab_dist
+            attn_dist_ = (1 - p_gen) * attn_dist
+
+            if extra_zeros is not None:
+                vocab_dist_ = paddle.concat([vocab_dist_, extra_zeros], 1)
+            final_dist = paddle2D_scatter_add(vocab_dist_, enc_batch_extend_vocab, attn_dist_, 1)
+        else:
+            final_dist = vocab_dist
+
+        return final_dist, s_t, c_t, attn_dist, p_gen, coverage
+
+
+class Model(object):
```

Comment:
Why not inherit from `paddle.nn.Layer`?

Reply:
This is one detail that we decided to follow the competitor (Pytorch). The Pytorch code sets up Model as an Object and optimizes the conjunction of the model.encoder.parameters+model.decoder.parameters+... instead of model.parameters, such that the shared embedding.weight between the encoder and the decoder is optimized twice.